Add transaction expiry slot for pending transactions

rvl opened the pull request:
### Issue Number

Relates to ADP-93 / #1838.

### Overview

- Adds expiry slot to pending transactions in order to implement transaction TTL.
- Expiry slot and time is reported in the transaction history.
- At the expiry slot, pending transactions are marked Expired.
- Default TTL of 7200 slots is still hard coded.

rvl committed:
Add transaction expiry slot for pending transactions

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -916,6 +919,25 @@ instance FromText TxStatus where
+    deriving (Generic, Eq, Show)
+
+instance Ord PendingTx where
+    compare (PendingTx a _) (PendingTx b _) = compare a b
```
```suggestion
    compare = comparing pendingTx
```

(Where `comparing` comes from `Data.Ord`.)

rvl replied:
Thanks - good idea.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -983,6 +1005,12 @@ fromTransactionInfo info = Tx
     }
 
+fromTransactionInfoPending :: TransactionInfo -> PendingTx
+fromTransactionInfoPending info = PendingTx
```
Perhaps the `xFromY` naming scheme would be more understandable here? So:

```suggestion
pendingTxFromTransactionInfo :: TransactionInfo -> PendingTx
pendingTxFromTransactionInfo info = PendingTx
```

rvl committed:
swagger.yaml: Add transaction expires_at and expired status

rvl committed:
tests: add TRANS_CREATE_10 - Pending transaction expiry

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -593,6 +594,36 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
+            , expectField #expiresAt (`shouldSatisfy` isJust)
+            ]
+
+        -- This stuff would be easier with Control.Lens...
```
Are we tracking that in a technical debt ticket :thinking: ?

rvl replied:
No, because I sent a message about this to slack, and got exactly 0 replies.

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -593,6 +594,36 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
+        let sl = sinceBlock ^. #absoluteSlotNumber . #getApiT
+
+        -- The expected expiry slot (adds the hardcoded default ttl)
+        let ttl = sl + SlotNo 7200
```
Should we perhaps expose that default "magic number" ? Or at least, have it behind a named constant in the integration DSL so that we do not duplicate it all over the place.

rvl replied:
OK done.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -840,6 +840,7 @@ restoreBlocks ctx wid blocks nodeTip = db & \DBLayer{..} -> mapExceptT atomicall
     let k = gp ^. #getEpochStability
     let localTip = currentTip $ NE.last cps
 
+    updatePendingTx (PrimaryKey wid) (view #slotNo localTip)
```
`update` feels a bit general here. When reading this, my first thought is "update what?". What about using a more specific qualifier like: "discardNowExpiredTx" or, whatever you'd come up with ?

rvl replied:
`updatePendingTxForExpiry`

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1332,9 +1333,11 @@ mkApiTransactionFromInfo ti (TransactionInfo txid ins outs ws meta depth txtime
         case meta ^. #status of
             Pending  -> #pendingSince
             InLedger -> #insertedAt
+            Expired  -> #pendingSince
```
Shall we perhaps use yet-another-field for tracking the time since which the transaction has expired?

rvl replied:
I think "pending_since" is the correct field. The time at which the transaction expired is in "expires_at".
So both times are captured for expired transactions.

KtorZ replied:
Fair enough!

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -663,6 +665,11 @@ data ApiBlockReference = ApiBlockReference
+data ApiSlotReference = ApiSlotReference
+    { time :: !UTCTime
+    , absoluteSlotNumber :: !(ApiT SlotNo)
+    } deriving (Eq, Generic, Show)
```
What about including also the epoch and slot within epoch to be consistent with other time references? We use:

```hs
data ApiTimeReference = ApiTimeReference
    { time :: !UTCTime
    , block :: !ApiBlockReference
    } deriving (Eq, Generic, Show)

data ApiBlockReference = ApiBlockReference
    { epochNumber :: !(ApiT EpochNo)
    , slotNumber :: !(ApiT SlotInEpoch)
    , height :: !(Quantity "block" Natural)
    , absoluteSlotNumber :: !(ApiT SlotNo)
    } deriving (Eq, Generic, Show)
```

Although indeed, the TTL is about slots and not blocks so the _height_ don't make much sense. The other however do. So I'd suggest to have something along the lines of:

```hs
data ApiTimeReferenceBySlot = ApiTimeReferenceBySlot
    { time :: !UTCTime
    , slot :: !ApiSlotReference
    } deriving (Eq, Generic, Show)

data ApiSlotReference = ApiBlockReference
    { epochNumber :: !(ApiT EpochNo)
    , slotNumber :: !(ApiT SlotInEpoch)
    , absoluteSlotNumber :: !(ApiT SlotNo)
    } deriving (Eq, Generic, Show)
```

rvl replied:
Not only is TTL about slots, but it's about slots in the _future_ -- which cannot have a block height.

OK I have added epoch and slot numbers.

KtorZ replied:
For consistency, It'd be better to have them nested under a `slot` field, don't you think :thinking: ?

rvl replied:
Yeah maybe - I am going to straighten out the types a bit (without changing existing API fields).

rvl committed:
Rename ApiTimeReference -> ApiTimeReferenceWithBlock

rvl committed:
Add epoch and local slot index to tx expiry slot reference

rvl committed:
Rename updatePendingTx to updatePendingTxForExpiry

rvl committed:
Add default TTL constant to tests DSL

KtorZ commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -522,11 +526,20 @@ walletId =
 
+-- | Wallet server's chosen transaction TTL value (in slots) when none is given.
+defaultTxTTL :: SlotNo
+defaultTxTTL = 7200
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/MVar.hs:
```diff
@@ -174,6 +175,9 @@ newDBLayer timeInterpreter = do
         -----------------------------------------------------------------------}
 
+        , updatePendingTxForExpiry = \pk tip -> ExceptT $ do
+            alterDB errNoSuchWallet db (mUpdatePendingTx pk tip)
```
:+1:

rvl committed:
Straighten out slot and block references in API

ApiBlockReference is just ApiSlotReference with block information
added -- block references gain slot info.

ApiNetworkTip is replaced with ApiSlotReference -- so it gains a
"time" field.

The Haskell API types have SlotId and block info nested rather than
inline. I hope in future that we can change the swagger spec to
match. But for now there are some non-generic Aeson instances to
maintain backwards compatibility.

rvl committed:
Update golden API files because the time field is added

KtorZ reviewed (approved):
Nice work!
